multicluster/controllers/multicluster/gateway_controller.go

```diff
@@ -0,0 +1,317 @@
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	r.leaderNamespace = remoteCluster.GetNamespace()
+
+	err = r.getServiceCIDR(ctx)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	resExportName := r.localClusterID + "-" + r.namespace + "-clusterinfo"
+	resExportNamespaced := types.NamespacedName{
+		Name:      resExportName,
+		Namespace: r.leaderNamespace,
+	}
+	resExport := &mcsv1alpha1.ResourceExport{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      resExportName,
+			Namespace: r.leaderNamespace,
+		},
+	}
+
+	createOrUpdate := func(gwIP string) error {
+		existResExport := &mcsv1alpha1.ResourceExport{}
+		if err := remoteCluster.Get(ctx, resExportNamespaced, existResExport); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return err
+			}
+			if err = r.createorUpdateResourceExport(ctx, req, remoteCluster, existResExport, gwIP, true); err != nil {
+				return err
+			}
+			return nil
+		}
+		if err = r.createorUpdateResourceExport(ctx, req, remoteCluster, existResExport, gwIP, false); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	gwn := &mcsv1alpha1.Gateway{}
+	if err := r.Client.Get(ctx, req.NamespacedName, gwn); err != nil {
+		if apierrors.IsNotFound(err) {
+			gwNodeInfos, err := r.getGatewayInfos()
+			if err != nil {
+				klog.ErrorS(err, "Failed to get Gateways")
+				return ctrl.Result{}, err
+			}
+			if len(gwNodeInfos) == 0 {
+				// When the last Gateway is deleted, we will remove the ClusterInfo kind of ResourceExport
+				if err := remoteCluster.Delete(ctx, resExport, &client.DeleteOptions{}); err != nil {
+					return ctrl.Result{}, client.IgnoreNotFound(err)
+				}
+				return ctrl.Result{}, nil
+			}
+			// When the deleted Gateway is not the last Gateway, we should create or update existing ResourceExport
+			// with the latest Gateway in remaining Gateways.
+			if err := createOrUpdate(gwNodeInfos[0].GatewayIP); err != nil {
+				return ctrl.Result{}, err
+			}
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+	if err := createOrUpdate(gwn.GatewayIP); err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *GatewayReconciler) getGatewayInfos() ([]mcsv1alpha1.GatewayInfo, error) {
+	gws := &mcsv1alpha1.GatewayList{}
+	var gatewaysInfo []mcsv1alpha1.GatewayInfo
+	if err := r.Client.List(ctx, gws, &client.ListOptions{}); err != nil {
+		return nil, err
+	}
+	if len(gws.Items) == 0 {
+		return nil, nil
+	}
+
+	// Sort Gateways by CreationTimestamp, the last created Gateway will be the first element.
+	sort.Slice(gws.Items, func(i, j int) bool {
+		return !gws.Items[i].CreationTimestamp.Before(&gws.Items[j].CreationTimestamp)
+	})
+
+	for _, n := range gws.Items {
+		gatewaysInfo = append(gatewaysInfo, mcsv1alpha1.GatewayInfo{
+			GatewayIP: n.GatewayIP,
+		})
+	}
+
+	// Make sure we only return the last created Gateway for now.
+	return []mcsv1alpha1.GatewayInfo{gatewaysInfo[0]}, nil
```

Comment:
Then why not return a single Gateway and name the func getLastCreatedGateway()? We can refactor later when needed.

Reply:
done